src/librpmem/rpmem_fip.c

```diff
@@ -943,29 +973,145 @@ rpmem_fip_fini_lanes_apm(struct rpmem_fip *fip)
 }
 
 /*
- * rpmem_fip_persist_raw -- (internal) perform persist operation using
- * READ after WRITE mechanism
+ * rpmem_fip_tx_inc -- (internal) increment number of elements in TX
+ */
+static inline void
+rpmem_fip_tx_inc(struct rpmem_fip_plane *lanep)
+{
+	++lanep->base.tx_elems;
+}
+
+/*
+ * rpmem_fip_tx_set_empty -- (internal) zero number of elements in TX
+ */
+static inline void
+rpmem_fip_tx_set_empty(struct rpmem_fip_plane *lanep)
+{
+	RPMEM_ASSERT(!lanep->base.tx_is_flushing);
+	lanep->base.tx_elems = 0;
+}
+
+/*
+ * rpmem_fip_tx_require_flush -- (internal) is TX almost full
+ */
+static inline int
+rpmem_fip_tx_require_flush(struct rpmem_fip *fip, struct rpmem_fip_plane *lanep)
+{
+	RPMEM_ASSERT(lanep->base.tx_elems < fip->fi->tx_attr->size);
+	return lanep->base.tx_elems + 1 == fip->fi->tx_attr->size;
+}
+
+/*
+ * rpmem_fip_tx_is_flushing -- (internal) is TX flush started
+ */
+static inline int
+rpmem_fip_tx_is_flushing(struct rpmem_fip_plane *lanep)
+{
+	return lanep->base.tx_is_flushing;
+}
+
+/*
+ * rpmem_fip_tx_set_flushing -- (internal) mark TX flush start
+ */
+static inline void
+rpmem_fip_tx_set_flushing(struct rpmem_fip_plane *lanep)
+{
+	lanep->base.tx_is_flushing = 1;
+}
+
+/*
+ * if TX is almost full last WRITE has to report its completion
+ * otherwise it is unknown when subsequent commands can be posted
+ */
+#define RPMEM_FIP_TX_FLUSH RPMEM_COMPLETION
+
+/*
+ * rpmem_fip_tx_flush_wait -- (internal) wait for WRITE completion
+ * to make sure TX can accept subsequent commands
+ */
+static inline int
+rpmem_fip_tx_flush_wait(struct rpmem_fip *fip, struct rpmem_fip_plane *lanep)
+{
+	RPMEM_ASSERT(lanep->base.tx_elems == fip->fi->tx_attr->size);
+	RPMEM_ASSERT(lanep->base.tx_is_flushing);
+
+	/* wait for WRITE completion */
+	int ret = rpmem_fip_lane_wait(fip, &lanep->base, FI_WRITE);
+	if (unlikely(ret)) {
+		ERR("waiting for WRITE completion failed");
+		return ret;
+	}
+	/* when WRITE completion is reaped TX is empty */
+	lanep->base.tx_is_flushing = 0;
```

Comment:
what about rpmem_fip_tx_set_flushingrpmem_fip_tx_set_flushing(struct rpmem_fip_plane *lanep, int value)

Reply:
I have assumed all TX queue related structures are modified inside dedicated ```rpmem_fip_tx_*``` functions. Functions outside from this set require the ability to set TX queue is flushing but set it off is only inside ```rpmem_fip_tx_flush_wait``` which is the correct way to wait for TX queue flush end.